# JSON and CSVs

## On boilerplate code

We are finally done with learning new syntax and structures in python.  Now we can start using all the patterns we've been learning in the context of real data.  

You'll be learning some new packages and have your first real interaction with boilerplate code.

Even though you've learned a lot these past few weeks, there are still structures and needs that some tools have that won't really make sense at first interaction.  Some tools are just designed to work in a certain way, and you are provided with documentation and example code that accomplishes them.  This is what we call boilerplate code.  

Much like how a syllabus is required to have statements about inclusion and diversity, not all professors are experts in the appropriate language to say for these things.  So we look to those relevant departments for guidance.  They often provide expertly written language for these documents to help ensure standards are met for all students and to avoid a misinterpretation of the policies.

These are often things that we read through and they make sense, but we often think "there's no way I could have written that on my own!" And that's kind of the point!  That's boilerplate code and that's boilerplate language.  For times when things are so complex and nuanced that the template is just given to you.  As your expertise in that area develops, you are better able to understand the inner workings and the design choices.

A lot of code is a bit of boilerplate that you'll need to adapt to fit your needs, but you'll need to leave the rest of it alone.  A bit part of the challenge is that you'll need to be able to look at the boilerplate code and make an educated guess about what parts of the code you need to change and what parts you should leave alone.  The trick is to find the input and the output of that chunk of code.  Where is the data going into it and where is it coming out.

I'm going to be giving you some boilerplate code for reading in and writing data of these two formats, but there are also some boilerplate patterns that are just the patterns that we use to accomplish these things. 

## What are json and csv?

### From a data structure perspective

These are less two different data types are more like two different data structures.  You can have many kinds of data within them, but they are two different ways of organizing data.

Much like having two different ways of organizing a kitchen or closet, each will get the job done so long as you know where things are, but you may find that:

* one structure makes more sense to you and is your preferred way of organizing the information
* the types of data and questions that you have are more easily processed in one structure than another
* you can generally accomplish the same things in each structure, but the different organization models will be better equipped to store certain data relationships than the other.

### From a format perspective

Each of these types of files are stored in a plain text file.  This means that you should be able to open it up and view it in any plain text editor, but you may not be able to really make sense of the structure just by looking at it.  That's because this format wasn't necessarily designed to organize data for human eyes.  The data is formatted in plain text to make it easy for a computer to process those things.  It is text designed for a program to parse.

Think of a CSV file that you might open up in excel.  You can see the content in the raw file, but you may really need the rendering of the excel workbook to be able to cleanly scan the data.

JSON files are much more like this.  They are not designed for human eyes at all and require programmatic processing for everything to get anything done to it.  What it has sacrificed in terms of readability, it did nso to gain an exceptional amount of depth of detail in the kind of data content and relationships that it can explore.

### From a technical perspective

Each of these files are plain text, so we process them in python much the same way we are used to other files. We read them in as a string or open them up as an infile as normal, and then pass that information off to another part of the program that knows what to do with that kind of data structure.  That code then does the job of translating the data into a structure that we know how to work with in python, and then we can get to business with that resulting structure.

This process is often called parsing of the data.  While we could write our own code to parse the content into a structure we can work with, we usually leave it up to python to deal with for us because we have better things to do with our time.

#### CSV structures in python

CSV documents are rectangular structures that are often best represented as 2D lists.  This means that it creates an outer list that holds many inner lists. Each inner list corresponds to a specific row in the data, and the elements within the list are the values for each cell going across that row.  Lists are a great way to store this data because tho position of the data within the file determines the meaning of it.  Lists have a primary access and identification structure around positions.  However, the thing that isn't well represented within a 2D list is the idea of the column.  You can easily grab an individual row from a 2D list, because that's just one of the inner lists.  We can use normal slicing syntax to grab an entire row of data (`list[pos_num]`).

However, columns in CSVs aren't all contained in a single data structure.  The values that correspond to a column all live across all the inner lists.  Thus, to get the values from an entire column we must first access every inner list and then grab that column position number.  There are special tools (like pandas) that were created to store this kind of 2D array of data in a way that we can access things like columns in a more normal fashion.  From a normal perspective you'd think that accessing a row and a column are about the same, but in the reality of the code, accessing data from a column looks completely different and much more complex.

There are tools you can use to read a CSV into an python `dict` structure, although we will not be using that here.

#### JSON structures in python

json has a completely different structure from a rectangular array of data like a CSV.  In fact, json has no specific shape this this that we can call it, other than a tree structure.  Whereas CSVs are very position first and everything is called via that position value, json has a very name or label oriented structure.

Operationally, you've been dealing with this kind of structure already with dictionaries in Python.  In fact, when you parse a json file into python, it is saved as either one big dictionary or a list of many dictionaries in there.  There are no rows of data, but instead clusters of trees or one large tree that you can access.  Even if those dictionaries are stored within a list, that doesn't mean that the position values for those dictionaries have an actual meaning to the data.  You may be able to infer the meaning based on your understanding of how the data file was created, but that won't always hold true.

## How to parse these in python

You're going to use boilerplate code for each of these to read them in.  There are modules in the Python standard library for parsing json and csv data.  We'll be using both of those.

Much of the credit for the patterns in use for this lesson are from The Python Cookbook 3rd edition.

# The data we'll be using

Take a moment to look through the `lunaresults.json` file in this directory.  It contains the wikipedia api results of 

# Parsing json with the `json` library

Here's the boilerplate code:

You'll see that you'll be reading in the json file like a normal text file with `.read()` so it is all one string.  Then passing that string into a special function within the library called `json.loads()` This function will parse a string and return back the relevant python data structure.

``` python
import json 

infile = open('yourjsonfile.json', 'r')
text = infile.read()
infile.close()

data = json.loads(text)
```

Your `data` variable is now a full python object.  json can actulaly have many structures internally, so you may have a dictionary or a list.

The first thing you should do is inspect your data variable and learn some things about it.  Only that can tell you what you should be doing with it.  

Also, there's now nothing particularly special about this variable.  The data is stored in pure python data structures.  You'll find dictionaries, lists, strings, numbers, etc within it.

In [1]:
import json 

infile = open('lunaresults.json', 'r')
text = infile.read()
infile.close()

data = json.loads(text)

In [3]:
print(type(data))

<class 'dict'>


We can see that this is a dictionary.  What are the keys?

In [4]:
print(data.keys())

dict_keys(['batchcomplete', 'query'])


Let's take a look inside the data file and get a sense of the structure.

In [5]:
data

{'batchcomplete': '',
 'query': {'search': [{'ns': 0,
    'pageid': 391371,
    'size': 47836,
    'snippet': 'Xenophilius <span class="searchmatch">Lovegood</span>, is the editor-in-chief of The Quibbler, a strange magazine claiming supposedly untrue things <span class="searchmatch">Luna</span> and her father believe. When <span class="searchmatch">Luna</span> was',
    'timestamp': '2018-10-16T20:06:57Z',
    'title': "Dumbledore's Army",
    'wordcount': 4967},
   {'ns': 0,
    'pageid': 3931304,
    'size': 39384,
    'snippet': 'podcast host and narrator. She rose to prominence for her portrayal of <span class="searchmatch">Luna</span> <span class="searchmatch">Lovegood</span> in the Harry Potter film series, appearing in the last four films and',
    'timestamp': '2018-10-10T09:09:37Z',
    'title': 'Evanna Lynch',
    'wordcount': 3060},
   {'ns': 0,
    'pageid': 874930,
    'size': 86965,
    'snippet': 'semi-humans, greatly attractive Wood Nymphs <span class="searchmatch">Lun

We can see that there's a bunch of data in there, but nested within some key structures.

Now you can see what is at the heart of json.  Storing things like a dictionary key/value pair structure means that you can store multiple data points and nest even more complex data structures within your data.

Let's look at the query key, because that's where our data lives.

This type of structure is completely normal to work with when it comes to apis.  Many times you are served up many data records all in one pile.  Being stored in json structures means that you can store a full and independent data record within a larger structure.

We need to start drilling down into query to get our data out.  Remember that this is a dictionary we're working with, so we need to use the dictionary style syntax to get things out of it.

In [7]:
data['query']

{'search': [{'ns': 0,
   'pageid': 391371,
   'size': 47836,
   'snippet': 'Xenophilius <span class="searchmatch">Lovegood</span>, is the editor-in-chief of The Quibbler, a strange magazine claiming supposedly untrue things <span class="searchmatch">Luna</span> and her father believe. When <span class="searchmatch">Luna</span> was',
   'timestamp': '2018-10-16T20:06:57Z',
   'title': "Dumbledore's Army",
   'wordcount': 4967},
  {'ns': 0,
   'pageid': 3931304,
   'size': 39384,
   'snippet': 'podcast host and narrator. She rose to prominence for her portrayal of <span class="searchmatch">Luna</span> <span class="searchmatch">Lovegood</span> in the Harry Potter film series, appearing in the last four films and',
   'timestamp': '2018-10-10T09:09:37Z',
   'title': 'Evanna Lynch',
   'wordcount': 3060},
  {'ns': 0,
   'pageid': 874930,
   'size': 86965,
   'snippet': 'semi-humans, greatly attractive Wood Nymphs <span class="searchmatch">Luna</span> <span class="searchmatch">Lovegood</span

Working with large dictionaries like this can be really overwhelming.  So we need to use a few more techniques to investigate it and get a feel for where out data is.

In [8]:
print(type(data['query']))

<class 'dict'>


So we see we have another dictionary.

In [9]:
for key, value in data['query'].items():
    print(key, type(value))

searchinfo <class 'dict'>
search <class 'list'>


This loop has gone over our keys and values and told us what the name of the key and the datatype is.  knowing the datatype is really important so we know how to access information from it.

Now that we are digging deeper, our dictionary syntax is going to get more complicated.  We can start chaining our key lookups.  But we'll want to save this next result as a variable!

Looks like the first key is searchinfo and it is holding some of the metadata about out api query.

In [10]:
data['query']['searchinfo']

{'totalhits': 55}

In this case, this small dictionary is telling us that there were 55 hits.  In the scope of this data file, this means that there should be 55 results waiting for us somewhere in this data structure.  We need to find something that has a length of 55.

Let's look at the search key and see what that list has.

In [11]:
data['query']['search']

[{'ns': 0,
  'pageid': 391371,
  'size': 47836,
  'snippet': 'Xenophilius <span class="searchmatch">Lovegood</span>, is the editor-in-chief of The Quibbler, a strange magazine claiming supposedly untrue things <span class="searchmatch">Luna</span> and her father believe. When <span class="searchmatch">Luna</span> was',
  'timestamp': '2018-10-16T20:06:57Z',
  'title': "Dumbledore's Army",
  'wordcount': 4967},
 {'ns': 0,
  'pageid': 3931304,
  'size': 39384,
  'snippet': 'podcast host and narrator. She rose to prominence for her portrayal of <span class="searchmatch">Luna</span> <span class="searchmatch">Lovegood</span> in the Harry Potter film series, appearing in the last four films and',
  'timestamp': '2018-10-10T09:09:37Z',
  'title': 'Evanna Lynch',
  'wordcount': 3060},
 {'ns': 0,
  'pageid': 874930,
  'size': 86965,
  'snippet': 'semi-humans, greatly attractive Wood Nymphs <span class="searchmatch">Luna</span> <span class="searchmatch">Lovegood</span>, a pupil at Hogwarts, and 

Hmm, that looks like a lot of stuff and we know that it's a list of data. So let's check the length.

In [12]:
len(data['query']['search'])

55

Yay! There's our 55!

Let's save this to a variable and call it records.

In [13]:
records = data['query']['search']

HOMEWORK HINT! The data structure that you'll be working with on your homework is a simplified version of this file and represents only the results coming in from this key.  This means that your data variable coming in from the json parser will be a list of dictionaries and not a dictionary like the start of this lecture. 

This means that you should start here if you are looking for more direct hints for how things should be working.

Now that we have a new data structure, let's repeat this process of inspection of what we have.

In [14]:
print(type(records))

<class 'list'>


In [15]:
print(len(records))

55


So we have a list of what?  Let's inspect the first one in this and see what's going on.  We can use regular slicing syntax here to get out the first element.

In [16]:
entry = records[0]

In [17]:
print(type(entry))
print(len(entry))

<class 'dict'>
7


So this is a dictionary of 7 key/value pairs.  Let's look at the keys and their datatypes.

In [19]:
for key, value in entry.items():
    print(key, type(value))

ns <class 'int'>
title <class 'str'>
pageid <class 'int'>
size <class 'int'>
wordcount <class 'int'>
snippet <class 'str'>
timestamp <class 'str'>


So all of these values are either a string of an int, which means that we don't have any more drilling down to do.

Let's check that all of the entries in this list are of type dictionary.  We're going to use a few things here.  First, we'll use a list accumulator to gather up all the data types.  Then we can do a few things to that list to inspect how many of what do we have.

In [20]:
types = [] 

for item in records:
    types.append(type(item))

In [22]:
print(types)

[<class 'dict'>, <class 'dict'>, <class 'dict'>, <class 'dict'>, <class 'dict'>, <class 'dict'>, <class 'dict'>, <class 'dict'>, <class 'dict'>, <class 'dict'>, <class 'dict'>, <class 'dict'>, <class 'dict'>, <class 'dict'>, <class 'dict'>, <class 'dict'>, <class 'dict'>, <class 'dict'>, <class 'dict'>, <class 'dict'>, <class 'dict'>, <class 'dict'>, <class 'dict'>, <class 'dict'>, <class 'dict'>, <class 'dict'>, <class 'dict'>, <class 'dict'>, <class 'dict'>, <class 'dict'>, <class 'dict'>, <class 'dict'>, <class 'dict'>, <class 'dict'>, <class 'dict'>, <class 'dict'>, <class 'dict'>, <class 'dict'>, <class 'dict'>, <class 'dict'>, <class 'dict'>, <class 'dict'>, <class 'dict'>, <class 'dict'>, <class 'dict'>, <class 'dict'>, <class 'dict'>, <class 'dict'>, <class 'dict'>, <class 'dict'>, <class 'dict'>, <class 'dict'>, <class 'dict'>, <class 'dict'>, <class 'dict'>]


There's a lot here, and we can try to eyeball it but we can also use python to look at stuff here.

What we really care about here are any unique values within the list.  There are a few ways we can do this.  First, we can try to look at the distinct values.  

The `set()` datatype in python will be good for this.  The set object is a little little like a list, where it will hold values.  This is actually a really fun mathematical tool that we can use, but set containers are really valuable for looking at combinations of data in lists.

Most important to know about sets is that they can only hold unique values.  This means that when you recast as list to a set it collapses all the values down to just the unique values.  So maybe you can see how handy this is?

Let's take a look.

In [23]:
print(set(types))

{<class 'dict'>}


We only have one result here, which means that all the values are the same across the entire original list.  This also means that all the items in the records list have the same data type.

We could use this same method to check how long each dictionary is within the list of records and check that they are uniform.  HOMEWORK HINT.

Let's now look at the words used in the snippet.  About how long is each snippet?  How uniform are they? Knowing that might influence what we choose to do with this text.  We're going to clean out all the punctuation from each and then split to get a rough approximation of how long each is.  This should give us a range.

While we're in there, we might as well count all the words, right?  Let's do that too since we already have the words split up.

We're going to use a nested for loop to first grab and split the text, and then the word count pattern to count all the words seen across all the records.

In [37]:

import string

def clean_line(line_str):
    cleanline = line_str
    for punc in string.punctuation:
        cleanline = cleanline.replace(punc, "")
    return cleanline

wordcount = {}
lengthcount = []

for item in records:
    snippet = item['snippet']
    clean_text = clean_line(snippet).lower()
    words = clean_text.split()
    # let's add the length of the words
    lengthcount.append(len(words))
    # now that we have the words we can loop over them and count them
    for word in words:
        if word in wordcount:
            wordcount[word] += 1
        else:
            wordcount[word] = 1

Let's take a look at the lengths of the snippets first and see how uniform the lengths might be.

In [39]:
print(lengthcount)

[26, 29, 27, 27, 29, 27, 28, 27, 26, 26, 30, 26, 28, 24, 28, 28, 27, 23, 27, 21, 29, 26, 25, 28, 29, 27, 26, 23, 26, 27, 26, 27, 26, 25, 27, 26, 27, 27, 27, 24, 28, 25, 29, 27, 25, 16, 24, 27, 23, 5, 20, 27, 25, 23, 24]


This looks pretty uniform, but remember that we can use set to see how many unique values there are.

In [40]:
print(set(lengthcount))

{5, 16, 20, 21, 23, 24, 25, 26, 27, 28, 29, 30}


Interesting!  There's one really short one.  Let's put a pin in that.

Now that we have all these counts, we can try and look at the total spread.  Let's look at the unique counts of this dictionary accumulator. This will give us a sense of how the data is spread around. 

We can get the list of values from this dictionary with the .values() method, then pass that into a set to see all the unique values.


In [34]:
uniquecounts = set(wordcount.values())
# let's sort it so we can see things better
print(uniquecounts)

{1, 2, 3, 4, 5, 6, 7, 8, 9, 10, 11, 12, 14, 15, 16, 17, 26, 27, 28, 32, 40, 57, 59, 83, 116}


Looks like the top word has 116 counts and the lowest is 1. Given that we have only 55 results, that's a word that must appear just more than twice for each snippet!

In [36]:
def sortbycount(pair):
    return pair[1]

pairs = list(wordcount.items())
pairs.sort(key = sortbycount, reverse = True)
pairs

[('span', 116),
 ('the', 83),
 ('classsearchmatchlovegoodspan', 59),
 ('classsearchmatchlunaspan', 57),
 ('and', 40),
 ('harry', 32),
 ('in', 28),
 ('of', 27),
 ('potter', 26),
 ('to', 17),
 ('a', 16),
 ('weasley', 15),
 ('lynch', 14),
 ('is', 12),
 ('evanna', 12),
 ('as', 11),
 ('who', 11),
 ('character', 10),
 ('her', 9),
 ('longbottom', 9),
 ('on', 9),
 ('from', 9),
 ('for', 8),
 ('neville', 8),
 ('by', 8),
 ('father', 7),
 ('ginny', 7),
 ('xenophilius', 6),
 ('hogwarts', 6),
 ('hermione', 6),
 ('was', 5),
 ('she', 5),
 ('series', 5),
 ('films', 5),
 ('at', 5),
 ('role', 5),
 ('he', 5),
 ('ron', 5),
 ('are', 5),
 ('but', 5),
 ('actress', 5),
 ('7', 5),
 ('quibbler', 4),
 ('played', 4),
 ('order', 4),
 ('phoenix', 4),
 ('has', 4),
 ('malfoy', 4),
 ('retrieved', 4),
 ('wand', 4),
 ('magazine', 3),
 ('film', 3),
 ('his', 3),
 ('editor', 3),
 ('classsearchmatchlovegoodspans', 3),
 ('that', 3),
 ('rowling', 3),
 ('dumbledores', 3),
 ('been', 3),
 ('cho', 3),
 ('chang', 3),
 ('draco', 3),

So maybe our interests have shifted and we want to create a data file that we can open up in excel and play around with.  We will need to create a CSV file for this.

# Creating CSV files

Now let's switch gears and think about reconfiguring this data into another format.  JSON is useful, but not something that we can easily open and inspect. Let's get this file into a rectangular format, like something that we can open in excel.

To do this, we need to change these records into things that resemble rows we might see in a spreadsheet.  These rows are stored as a list, with each element in the list representing the individual cell values within that row.  This means you're going to do a double list accumulator pattern. The outside accumulator will hold all the lists that represent the rows. The for loop will iterate over the records in your data and access all the desired values and assemble the lists for each row.  You will only need a single for loop here, as you can extract the values from the record dictionary directly without looping over the values.

We need to start with our desired data values.

Let's say we want the following headers for our csv:

pageid, snippetwordcount

Two of these values are coming in directly from our dictionaries, and one we can compute on the snippet during a loop.

Here's some psuedocode for how to do this:

```
# set your primary list accumulator base, called allrows, to be an empty list

# loop over all the data records (indented comments go in your for loop)

    # set your row accumulator base to an empty list (call the variable row)

    # extract all the values you need from each record dict, saving them to variable
    # names that match the content. 

    # use append to add each value you've extracted into the row accumulator

    # add an assert statement that the length of row should be 4. Set the error message to
    # say "The length of the row list isn't 4".

    # at the very end, append that row to your allrows list
```

I'm also going to move the word count stuff into a function, so we don't clutter up out for loop.

In [45]:
import string

def clean_line(line_str):
    cleanline = line_str
    for punc in string.punctuation:
        cleanline = cleanline.replace(punc, "")
    return cleanline

def count_words(bigstring):
    clean_text = clean_line(bigstring).lower()
    words = clean_text.split()
    return len(words)

Now we can loop through out records and grab the data.

In [46]:
allrows = []

for record in records:
    row = []
    pageid = record['pageid']
    snippetcount = count_words(record['snippet'])
    row.append(pageid)
    row.append(snippetcount)
    assert len(row) == 2, "The length of the row list isn't 2"
    allrows.append(row)

Let's write an assert statement that our results list is 55 items long.

In [48]:
assert len(allrows) == 55, "The len isn't 55, you're missing data!"

In [49]:
allrows

[[391371, 26],
 [3931304, 29],
 [874930, 27],
 [23483380, 27],
 [38290297, 29],
 [891112, 27],
 [11061022, 28],
 [3512033, 27],
 [670407, 26],
 [45100, 26],
 [667601, 30],
 [827110, 26],
 [10093585, 28],
 [2298654, 24],
 [241855, 28],
 [9834441, 28],
 [48543224, 27],
 [771537, 23],
 [271050, 27],
 [855389, 21],
 [32402649, 29],
 [3833378, 26],
 [5032848, 25],
 [34085574, 28],
 [44334, 29],
 [21419143, 27],
 [19156567, 26],
 [906492, 23],
 [59572, 26],
 [156489, 27],
 [156587, 26],
 [33976749, 27],
 [776049, 26],
 [667658, 25],
 [858575, 27],
 [284598, 26],
 [17763993, 27],
 [28373862, 27],
 [54465045, 27],
 [31941988, 24],
 [779883, 28],
 [27511982, 25],
 [32446375, 29],
 [12976982, 27],
 [3381404, 25],
 [12294600, 16],
 [24870495, 24],
 [20997070, 27],
 [29494722, 23],
 [29960484, 5],
 [45256977, 20],
 [43136825, 27],
 [947583, 25],
 [29494707, 23],
 [2228434, 24]]

Now that we have the list of rows, we can go through the process of writing out the CSV.

First, you need to make a list that contains the headers you want to appear in your file.  Each header should be a string with the text value.  Your list should be the same length as the internal lists for each row (so 2 in this case.)

In [50]:
headers = ['pageid', 'wordcount']

Now we can go through our boilerplate code for writing out the CSV.  This uses a different structure for opening a fileio object, but you still will see an outfile object.  Importantly, there's no need to close the outfile because this pattern takes care of that for you.

``` python
import csv

with open('youroutfilename.csv', 'w') as outfile:
    csvout = csv.writer(outfile)
    csvout.writerow(headers)
    csvout.writerows(allrows)
```

First we import the csv module.  This gives us specific functions for parsing CSVs, including writing them.

Then, we are using the with block keyword to open our file.  This structure will automatically close the file we have opened as soon as the with block is exited from.

This is one that you won't really understand right now, just note the things that you'll need to change.  Let's finish walking through it and then I'll tell you what you need to change.

There are three lines of code within this with statement.  We do have a regular outfile object, just opened differently.  This fileio object is passed into the `csv.writer()` function.  That function transforms the fileio object into a csvwriter object.  This is a special kind of io object that knows how to write csv data via special methods.  This is all you need to understand about how this is working

The next two lines are demonstrations of these special methods.  Each requires a different kind of data structure.

`.writerow(headers)` takes a single list of strings.  That row will be written as one line of the file, with the elements within it written separated by commas.  

`.writerows(allrows)` takes a list of lists.  Each of those sublists needs to have strings, and the length of each list need to be the same.  This will write out each sublist as a separate row, with each string in that list separated by a comma.

What you need to provide?

* the file name in the `open()` function
* the name of the outfile variable you want (outfile is usually fine)
* the csvout variable you want (usually csvoutfile is fine)
* the variable name with the headers list goes within `.writerow()`
* the variable name with the list of lists goes within `.writerows()`.

And that's it!  Execute the code and look at your file.

In [51]:
import csv

headers = ['pageid', 'snippetcount']

with open('lunasnippetcount.csv', 'w') as outfile:
    csvout = csv.writer(outfile)
    csvout.writerow(headers)
    csvout.writerows(allrows)

Now we can see that the pageid of 29960484 has the shortest snippet.  Let's find that in the data.

In [56]:
for entry in records:
    if entry['pageid'] == 29960484: # these are stored as ints in the dicts
        print(entry)

{'ns': 0, 'title': 'Bowie Wong', 'pageid': 29960484, 'size': 6402, 'wordcount': 768, 'snippet': 'au/fashion/wrap/vraw_livestyle2010 https://bowienews.wordpress.com/2012/06/04/from-<span class="searchmatch">luna</span>-<span class="searchmatch">lovegood</span>-to-looking-lovely-in-bowie/ http://www.stylebistro', 'timestamp': '2018-10-06T15:03:49Z'}


# Reading in csvs


We're roughly going to repeat the process of writing things out, creating really similar objects that we started with.  We'll read in the headers into a single list, then read in all the data rows as a list of lists.  You'll note that the boilerplate is pretty similar, but all the things have been flipped for file reading, and we have assignment statements along with our read methods.

You'll also see that we have a list accumulator happening to collect all the lines.  This is actually different from the standard boilerplate, as that would use something that you've never really seen before.

Here's the boilerplate.

``` python
import csv

allrows = []

with open('yourexistingfilename.csv', 'r') as infile:
    csvin = csv.reader(infile)
    headers = next(csvin)
    for line in csvin:
        allrows.append(line)
```

In [4]:
import csv

lunadata = []

with open('lunasnippetcount.csv', 'r') as infile:
    csvin = csv.reader(infile)
    headers = next(csvin)
    for line in csvin:
        lunadata.append(line)

Now we can inspect our structures and see what we have.

In [5]:
headers

['pageid', 'snippetcount']

In [7]:
print(len(lunadata))

55


Now we can us all our normal stuff for interacting with this to get out data and look at values.  Let's use this structure to sort our values so we can see the pageids with the pagecount.  We're going to use the sorting method that we learned last week.

In [11]:
def getcount(row):
    return row[1]

lunadata.sort(key = getcount)

lunadata

[['12294600', '16'],
 ['45256977', '20'],
 ['855389', '21'],
 ['771537', '23'],
 ['906492', '23'],
 ['29494722', '23'],
 ['29494707', '23'],
 ['2298654', '24'],
 ['31941988', '24'],
 ['24870495', '24'],
 ['2228434', '24'],
 ['5032848', '25'],
 ['667658', '25'],
 ['27511982', '25'],
 ['3381404', '25'],
 ['947583', '25'],
 ['391371', '26'],
 ['670407', '26'],
 ['45100', '26'],
 ['827110', '26'],
 ['3833378', '26'],
 ['19156567', '26'],
 ['59572', '26'],
 ['156587', '26'],
 ['776049', '26'],
 ['284598', '26'],
 ['874930', '27'],
 ['23483380', '27'],
 ['891112', '27'],
 ['3512033', '27'],
 ['48543224', '27'],
 ['271050', '27'],
 ['21419143', '27'],
 ['156489', '27'],
 ['33976749', '27'],
 ['858575', '27'],
 ['17763993', '27'],
 ['28373862', '27'],
 ['54465045', '27'],
 ['12976982', '27'],
 ['20997070', '27'],
 ['43136825', '27'],
 ['11061022', '28'],
 ['10093585', '28'],
 ['241855', '28'],
 ['9834441', '28'],
 ['34085574', '28'],
 ['779883', '28'],
 ['3931304', '29'],
 ['38290297', '29'],


That didn't seem to do the trick.  What happened?  Look at how this is sorting.  This is sorting by ascii value and not integer.  We don't actually need to change the values in our data to fix this, we can return an int value in the getcount function to have it sort by the int value.  Let's make that change and take a look.

In [12]:
def getcount(row):
    return int(row[1])

lunadata.sort(key = getcount)

lunadata

[['29960484', '5'],
 ['12294600', '16'],
 ['45256977', '20'],
 ['855389', '21'],
 ['771537', '23'],
 ['906492', '23'],
 ['29494722', '23'],
 ['29494707', '23'],
 ['2298654', '24'],
 ['31941988', '24'],
 ['24870495', '24'],
 ['2228434', '24'],
 ['5032848', '25'],
 ['667658', '25'],
 ['27511982', '25'],
 ['3381404', '25'],
 ['947583', '25'],
 ['391371', '26'],
 ['670407', '26'],
 ['45100', '26'],
 ['827110', '26'],
 ['3833378', '26'],
 ['19156567', '26'],
 ['59572', '26'],
 ['156587', '26'],
 ['776049', '26'],
 ['284598', '26'],
 ['874930', '27'],
 ['23483380', '27'],
 ['891112', '27'],
 ['3512033', '27'],
 ['48543224', '27'],
 ['271050', '27'],
 ['21419143', '27'],
 ['156489', '27'],
 ['33976749', '27'],
 ['858575', '27'],
 ['17763993', '27'],
 ['28373862', '27'],
 ['54465045', '27'],
 ['12976982', '27'],
 ['20997070', '27'],
 ['43136825', '27'],
 ['11061022', '28'],
 ['10093585', '28'],
 ['241855', '28'],
 ['9834441', '28'],
 ['34085574', '28'],
 ['779883', '28'],
 ['3931304', '29'],
 